File: frontend/src/lib/workflow/validation.ts
import { getAllowedBranchesForNode } from "./branches";
import { blocksByType as fallbackBlocksByType } from "./catalog";
import type { BlockLookup } from "./branches";
import type { BlockParam, ValidationIssue, ValidationResult, WorkflowPayload } from "./types";

export function validateWorkflowClient(
  payload: WorkflowPayload,
  blocksByType: BlockLookup = fallbackBlocksByType
): ValidationResult {
  const errors: ValidationIssue[] = [];
  const warnings: ValidationIssue[] = [];
  const { nodes, edges } = payload.definition;
  const nodeIdList = nodes.map((node) => node.id);
  const nodeIds = new Set(nodeIdList);
  const nodesById = new Map(nodes.map((node) => [node.id, node]));

  if (!nodes.length) {
    return {
      valid: false,
      errors: [{ code: "empty_workflow", message: "The workflow needs at least one node." }],
      warnings
    };
  }

  if (nodeIdList.length !== nodeIds.size) {
    errors.push({
      code: "duplicate_node_id",
      message: "Every workflow node must have a unique id."
    });
  }

  for (const node of nodes) {
    if (!blocksByType[node.type]) {
      errors.push({
        code: "unknown_node_type",
        message: `Unknown node type '${node.type}'.`,
        nodeId: node.id
      });
    }
  }

  const startNodes = nodes.filter((node) => node.type === "campaign_entrypoint");
  if (startNodes.length !== 1) {
    errors.push({
      code: "invalid_start_count",
      message: "The workflow must contain exactly one Campaign Start node."
    });
  }

  if (!nodes.some((node) => blocksByType[node.type]?.terminal)) {
    errors.push({
      code: "missing_terminal_node",
      message: "At least one risk end-state node is required."
    });
  }

  const outgoing = new Map<string, string[]>();
  const incoming = new Map<string, string[]>();
  const branchBySource = new Map<string, string[]>();
  for (const edge of edges) {
    if (!nodeIds.has(edge.source)) {
      errors.push({
        code: "missing_edge_source",
        message: "An edge references a missing source node.",
        edgeId: edge.id
      });
      continue;
    }

    if (!nodeIds.has(edge.target)) {
      errors.push({
        code: "missing_edge_target",
        message: "An edge references a missing target node.",
        edgeId: edge.id
      });
      continue;
    }

    outgoing.set(edge.source, [...(outgoing.get(edge.source) ?? []), edge.target]);
    incoming.set(edge.target, [...(incoming.get(edge.target) ?? []), edge.source]);
    branchBySource.set(edge.source, [...(branchBySource.get(edge.source) ?? []), edge.branch]);

    const sourceNode = nodesById.get(edge.source);
    const sourceBlock = sourceNode ? blocksByType[sourceNode.type] : null;
    const allowedBranches = getAllowedBranchesForNode(sourceNode, blocksByType);

    if (allowedBranches.length && !allowedBranches.includes(edge.branch)) {
      errors.push({
        code: "invalid_branch",
        message: `Branch '${edge.branch}' is not valid for ${sourceNode?.label ?? edge.source}.`,
        nodeId: edge.source,
        edgeId: edge.id
      });
    }

    if (sourceBlock?.terminal) {
      errors.push({
        code: "terminal_node_has_output",
        message: "Risk end-state nodes cannot have outgoing edges.",
        nodeId: edge.source,
        edgeId: edge.id
      });
    }
  }

  for (const node of nodes) {
    const block = blocksByType[node.type];
    if (!block) {
      continue;
    }

    for (const param of block.params) {
      const value = node.params[param.name];
      if (isEmptyParamValue(value)) {
        if (!param.required) continue;
        errors.push({
          code: "missing_required_param",
          message: `'${param.label}' is required.`,
          nodeId: node.id
        });
        continue;
      }

      const paramIssue = validateParamValue(param, value, node.id);
      if (paramIssue) {
        errors.push(paramIssue);
      }
    }

    const branches = branchBySource.get(node.id) ?? [];
    if (new Set(branches).size !== branches.length) {
      errors.push({
        code: "duplicate_branch",
        message: "A node cannot use the same branch label more than once.",
        nodeId: node.id
      });
    }

    if (node.type === "condition") {
      const allowedBranches = getAllowedBranchesForNode(node, blocksByType);
      const missingBranches = allowedBranches.filter((branch) => !branches.includes(branch));
      if (allowedBranches.length && missingBranches.length) {
        errors.push({
          code: "condition_missing_outcome",
          message: `Condition is missing outcome branch: ${missingBranches.join(", ")}.`,
          nodeId: node.id
        });
      }
      if (!allowedBranches.length && branches.length < 2) {
        errors.push({
          code: "condition_needs_branches",
          message: "Condition nodes need at least two outgoing branches.",
          nodeId: node.id
        });
      }
    }

    if (node.type !== "condition" && !block.terminal && branches.length > 1) {
      warnings.push({
        code: "multi_output_action",
        message: "Action nodes usually have a single outgoing edge.",
        nodeId: node.id
      });
    }

    if (node.type !== "campaign_entrypoint" && !(incoming.get(node.id) ?? []).length) {
      warnings.push({
        code: "missing_incoming_edge",
        message: "This node is not connected from any previous step.",
        nodeId: node.id
      });
    }
  }

  if (startNodes.length) {
    const reachable = reachableFrom(startNodes[0].id, outgoing);
    for (const node of nodes) {
      if (!reachable.has(node.id)) {
        warnings.push({
          code: "unreachable_node",
          message: "This node is not reachable from Campaign Start.",
          nodeId: node.id
        });
      }
    }
  }

  if (hasCycle([...nodeIds], outgoing)) {
    errors.push({ code: "cycle_detected", message: "The workflow graph must be acyclic." });
  }

  return { valid: errors.length === 0, errors, warnings };
}

function validateParamValue(
  param: BlockParam,
  value: unknown,
  nodeId: string
): ValidationIssue | null {
  if (param.kind === "select") {
    if (typeof value !== "string" || (param.options?.length && !param.options.includes(value))) {
      return {
        code: "invalid_param_option",
        message: `'${param.label}' must be one of: ${(param.options ?? []).join(", ")}.`,
        nodeId
      };
    }
  }

  if (param.kind === "number" && (typeof value !== "number" || !Number.isFinite(value))) {
    return {
      code: "invalid_param_type",
      message: `'${param.label}' must be a finite number.`,
      nodeId
    };
  }

  if (param.kind === "text" && typeof value !== "string") {
    return {
      code: "invalid_param_type",
      message: `'${param.label}' must be text.`,
      nodeId
    };
  }

  return null;
}

function isEmptyParamValue(value: unknown) {
  return value === undefined || value === null || value === "";
}

function hasCycle(nodeIds: string[], outgoing: Map<string, string[]>) {
  const visiting = new Set<string>();
  const visited = new Set<string>();

  function visit(nodeId: string): boolean {
    if (visiting.has(nodeId)) return true;
    if (visited.has(nodeId)) return false;
    visiting.add(nodeId);
    for (const target of outgoing.get(nodeId) ?? []) {
      if (visit(target)) return true;
    }
    visiting.delete(nodeId);
    visited.add(nodeId);
    return false;
  }

  return nodeIds.some((nodeId) => visit(nodeId));
}

function reachableFrom(startId: string, outgoing: Map<string, string[]>) {
  const visited = new Set<string>();
  const queue = [startId];

  while (queue.length) {
    const nodeId = queue.shift();
    if (!nodeId || visited.has(nodeId)) continue;
    visited.add(nodeId);
    queue.push(...(outgoing.get(nodeId) ?? []));
  }

  return visited;
}

File: backend/app/services/validation.py
from collections import defaultdict, deque
from math import isfinite
from typing import Any

from app.schemas.workflow import (
    ValidationIssue,
    ValidationResult,
    WorkflowDefinition,
    WorkflowNode,
)
from app.services.blocks import get_static_block_catalog


def _blocks_by_type(block_catalog: list[dict[str, Any]]) -> dict[str, dict[str, Any]]:
    return {block["type"]: block for block in block_catalog}


def validate_workflow(
    definition: WorkflowDefinition,
    block_catalog: list[dict[str, Any]] | None = None,
) -> ValidationResult:
    errors: list[ValidationIssue] = []
    warnings: list[ValidationIssue] = []
    blocks_by_type = _blocks_by_type(block_catalog or get_static_block_catalog())
    terminal_node_types = {
        block_type
        for block_type, block in blocks_by_type.items()
        if block.get("terminal") is True
    }

    nodes = definition.nodes
    edges = definition.edges
    node_ids = [node.id for node in nodes]
    node_id_set = set(node_ids)
    nodes_by_id = {node.id: node for node in nodes}

    if not nodes:
        errors.append(
            ValidationIssue(
                code="empty_workflow",
                message="The workflow needs at least one node.",
            )
        )
        return ValidationResult(valid=False, errors=errors, warnings=warnings)

    if len(node_ids) != len(node_id_set):
        errors.append(
            ValidationIssue(
                code="duplicate_node_id",
                message="Every workflow node must have a unique id.",
            )
        )

    unknown_nodes = [node for node in nodes if node.type not in blocks_by_type]
    for node in unknown_nodes:
        errors.append(
            ValidationIssue(
                code="unknown_node_type",
                message=f"Unknown node type '{node.type}'.",
                nodeId=node.id,
            )
        )

    start_nodes = [node for node in nodes if node.type == "campaign_entrypoint"]
    if len(start_nodes) != 1:
        errors.append(
            ValidationIssue(
                code="invalid_start_count",
                message="The workflow must contain exactly one Campaign Start node.",
            )
        )

    terminal_nodes = [node for node in nodes if node.type in terminal_node_types]
    if not terminal_nodes:
        errors.append(
            ValidationIssue(
                code="missing_terminal_node",
                message="At least one risk end-state node is required.",
            )
        )

    outgoing: dict[str, list[str]] = defaultdict(list)
    incoming: dict[str, list[str]] = defaultdict(list)
    edge_branches_by_source: dict[str, list[str]] = defaultdict(list)

    for edge in edges:
        if edge.source not in node_id_set:
            errors.append(
                ValidationIssue(
                    code="missing_edge_source",
                    message="An edge references a missing source node.",
                    edgeId=edge.id,
                )
            )
            continue
        if edge.target not in node_id_set:
            errors.append(
                ValidationIssue(
                    code="missing_edge_target",
                    message="An edge references a missing target node.",
                    edgeId=edge.id,
                )
            )
            continue

        outgoing[edge.source].append(edge.target)
        incoming[edge.target].append(edge.source)
        edge_branches_by_source[edge.source].append(edge.branch)

        source_node = nodes_by_id[edge.source]
        block = blocks_by_type.get(source_node.type, {})
        allowed_branches = set(_allowed_branches_for_node(source_node, blocks_by_type))
        if allowed_branches and edge.branch not in allowed_branches:
            errors.append(
                ValidationIssue(
                    code="invalid_branch",
                    message=f"Branch '{edge.branch}' is not valid for {source_node.label}.",
                    nodeId=edge.source,
                    edgeId=edge.id,
                )
            )
        if block.get("terminal"):
            errors.append(
                ValidationIssue(
                    code="terminal_node_has_output",
                    message="Risk end-state nodes cannot have outgoing edges.",
                    nodeId=edge.source,
                    edgeId=edge.id,
                )
            )

    for node in nodes:
        block = blocks_by_type.get(node.type)
        if not block:
            continue

        for param in block.get("params", []):
            value = node.params.get(param["name"])
            if _is_empty_param_value(value):
                if not param.get("required"):
                    continue
                errors.append(
                    ValidationIssue(
                        code="missing_required_param",
                        message=f"'{param['label']}' is required.",
                        nodeId=node.id,
                    )
                )
                continue

            param_issue = _validate_param_value(param, value, node.id)
            if param_issue:
                errors.append(param_issue)

        branch_names = edge_branches_by_source[node.id]
        if len(branch_names) != len(set(branch_names)):
            errors.append(
                ValidationIssue(
                    code="duplicate_branch",
                    message="A node cannot use the same branch label more than once.",
                    nodeId=node.id,
                )
            )

        if node.type == "condition":
            allowed_branches = _allowed_branches_for_node(node, blocks_by_type)
            missing_branches = [
                branch for branch in allowed_branches if branch not in branch_names
            ]
            if allowed_branches and missing_branches:
                errors.append(
                    ValidationIssue(
                        code="condition_missing_outcome",
                        message=(
                            "Condition is missing outcome branch: "
                            f"{', '.join(missing_branches)}."
                        ),
                        nodeId=node.id,
                    )
                )
            if not allowed_branches and len(branch_names) < 2:
                errors.append(
                    ValidationIssue(
                        code="condition_needs_branches",
                        message="Condition nodes need at least two outgoing branches.",
                        nodeId=node.id,
                    )
                )

        if node.type != "condition" and not block.get("terminal") and len(branch_names) > 1:
            warnings.append(
                ValidationIssue(
                    code="multi_output_action",
                    message="Action nodes usually have a single outgoing edge.",
                    nodeId=node.id,
                )
            )

        if node.type != "campaign_entrypoint" and not incoming[node.id]:
            warnings.append(
                ValidationIssue(
                    code="missing_incoming_edge",
                    message="This node is not connected from any previous step.",
                    nodeId=node.id,
                )
            )

    if start_nodes:
        reachable = _reachable_from(start_nodes[0].id, outgoing)
        for node in nodes:
            if node.id not in reachable:
                warnings.append(
                    ValidationIssue(
                        code="unreachable_node",
                        message="This node is not reachable from Campaign Start.",
                        nodeId=node.id,
                    )
                )

    if _has_cycle(node_id_set, outgoing):
        errors.append(
            ValidationIssue(
                code="cycle_detected",
                message="The workflow graph must be acyclic.",
            )
        )

    return ValidationResult(valid=not errors, errors=errors, warnings=warnings)


def _validate_param_value(
    param: dict[str, Any],
    value: Any,
    node_id: str,
) -> ValidationIssue | None:
    kind = param.get("kind")
    label = param.get("label", param.get("name", "Parameter"))

    if kind == "select":
        options = param.get("options") or []
        if not isinstance(value, str) or (options and value not in options):
            return ValidationIssue(
                code="invalid_param_option",
                message=f"'{label}' must be one of: {', '.join(options)}.",
                nodeId=node_id,
            )

    if kind == "number":
        if isinstance(value, bool) or not isinstance(value, (int, float)) or not isfinite(value):
            return ValidationIssue(
                code="invalid_param_type",
                message=f"'{label}' must be a finite number.",
                nodeId=node_id,
            )

    if kind == "text" and not isinstance(value, str):
        return ValidationIssue(
            code="invalid_param_type",
            message=f"'{label}' must be text.",
            nodeId=node_id,
        )

    return None


def _allowed_branches_for_node(
    node: WorkflowNode,
    blocks_by_type: dict[str, dict[str, Any]],
) -> list[str]:
    block = blocks_by_type.get(node.type, {})
    branch_rule = block.get("branchRule")
    if branch_rule:
        value = node.params.get(branch_rule["param"])
        if not isinstance(value, str):
            return []
        return branch_rule.get("branchesByValue", {}).get(value, [])

    return block.get("allowedBranches", [])


def _is_empty_param_value(value: Any) -> bool:
    return value is None or value == ""


def _reachable_from(start_id: str, outgoing: dict[str, list[str]]) -> set[str]:
    visited: set[str] = set()
    queue: deque[str] = deque([start_id])
    while queue:
        node_id = queue.popleft()
        if node_id in visited:
            continue
        visited.add(node_id)
        queue.extend(outgoing[node_id])
    return visited


def _has_cycle(node_ids: set[str], outgoing: dict[str, list[str]]) -> bool:
    visiting: set[str] = set()
    visited: set[str] = set()

    def visit(node_id: str) -> bool:
        if node_id in visiting:
            return True
        if node_id in visited:
            return False

        visiting.add(node_id)
        for target_id in outgoing[node_id]:
            if visit(target_id):
                return True
        visiting.remove(node_id)
        visited.add(node_id)
        return False

    return any(visit(node_id) for node_id in node_ids if node_id not in visited)
